packages/core/src/tools/boundedBuffer.ts

```diff
@@ -1,17 +1,24 @@
 const BUFFER_LIMIT = 500
 
-export class BoundedBuffer {
-  private buffer: Array<() => void> = []
+export class BoundedBuffer<T = void> {
+  private buffer: Array<(arg: T) => void> = []
 
-  add(callback: () => void) {
+  add(callback: (arg: T) => void) {
     const length = this.buffer.push(callback)
     if (length > BUFFER_LIMIT) {
       this.buffer.splice(0, 1)
     }
   }
 
-  drain() {
-    this.buffer.forEach((callback) => callback())
+  remove(callback: (arg: T) => void) {
+    const index = this.buffer.indexOf(callback)
+    if (index !== -1) {
+      this.buffer.splice(index, 1)
+    }
```

Comment:
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**🥜 nitpick:** ‏we are doing that at a couple of other places in the code, what about extracting a method in `arrayUtils`?